jupyter/swan.py

```diff
@@ -0,0 +1,638 @@
+
+    Instance can be used as function to refere to new name.
+    """
+
+    def __init__(self, columns_to_rename):
+        self.columns_to_rename = columns_to_rename
+
+    def rename(self, df):
+        """ Rename columns according self.columns_to_rename.
+            Use Renamer instance() method to refer to new names.
+
+        :returns: new dataframe with columns renamed
+
+        """
+        return df.rename(columns=self.columns_to_rename)
+
+    def __call__(self, original_name):
+        """ Returns new name of column to be used by formatting & styling functions. """
+        return self.columns_to_rename.get(original_name, original_name)
+
+
+def add_extra_and_composite_columns(df, slo):
+    """ Add extra derived columns with achieved normalized QPS/latency
+    and composite column to store all values in one dict.
+
+    Reshaping and preparing extra normalized columns
+    for latency/qps according target SLOs.
+
+    :returns: New dataframe with new derived columns and one special composite column.
+    """
+    # Extra columns.
+    # Calculate achieved QPS as percentage (normalized to 1).
+    df[ACHIEVED_QPS_LABEL] = pd.Series(df[QPS_LABEL] / df[SWAN_LOAD_POINT_QPS_LABEL])
+
+    # Calculate achieved latency in regards to SLO.
+    df[ACHIEVED_LATENCY_LABEL] = pd.Series(df[PERCENTILE99TH_LABEL] / slo)
+
+    # Columns to store in one cell for father processing.
+    COMPOSITE_COLUMNS = [ACHIEVED_QPS_LABEL, PERCENTILE99TH_LABEL, ACHIEVED_LATENCY_LABEL, QPS_LABEL]
+    # Composite value to store all values e.g. "achieved qps" and "latency" together in one cell as dict.
+    # Used to display one of the values and format using other value.
+    df[COMPOSITE_VALUES_LABEL] = df[COMPOSITE_COLUMNS].apply(dict, axis=1)
+
+    return df
+
+
+class Experiment:
+    """ Base class for loading & storing data for swan experiments.
+
+    Extra normalized columns are then stored in one composite column for further processing.
+
+    Additionally provides a helper function to rename existing dataframe columns to new names and
+    function that allows to refere to new names using original names (_renamed).
+    """
+
+    def __init__(self, experiment_id, cassandra_options=None, aggfunc=np.mean):
+        self.df = load_dataframe_from_cassandra(experiment_id, cassandra_options, aggfunc=aggfunc)
+        self.experiment_id = experiment_id
+        self.df.columns.name = 'Experiment %s' % self.experiment_id
+
+    def _repr_html_(self):
+        return self.df._repr_html_()
+
+
+class Profile:
+    """ Visualization for "sensitivity profile" experiments that presents
+        latency/QPS and caffe aggressor throughput in "aggressor" and
+        "load" dimensions.
+    """
+
+    def __init__(self, experiment, slo):
+        self.experiment = experiment
+        self.slo = slo
+
+        # Pre-process data specifically for this experiment.
+        df = self.experiment.df.copy()
```

Comment:
it is not - it the sole responsbility of Experiment class to provide data in "df"